File: sdk/arduino/IoTDevice/IoTDevice.cpp
/**
 * IoT Platform 设备接入 SDK — Arduino 实现
 */

#include "IoTDevice.h"

IoTDevice* IoTDevice::_instance = nullptr;

IoTDevice::IoTDevice(const char* deviceId, const char* broker, int port)
    : _deviceId(deviceId), _broker(broker), _port(port),
      _mqttClient(_wifiClient), _commandHandler(nullptr),
      _connected(false), _lastReconnectAttempt(0) {
    _instance = this;
}

bool IoTDevice::connect() {
    _mqttClient.setServer(_broker, _port);
    _mqttClient.setCallback(_mqttCallback);
    _mqttClient.setBufferSize(512);

    // 设置 LWT
    String willTopic = _topic("status");
    String willPayload = "{\"status\":\"offline\",\"reason\":\"connection_lost\"}";

    if (_mqttClient.connect(_deviceId, NULL, NULL,
                            willTopic.c_str(), 1, true,
                            willPayload.c_str())) {
        // 订阅命令主题
        String cmdTopic = _topic("commands");
        _mqttClient.subscribe(cmdTopic.c_str());

        // 发布上线状态
        String statusTopic = _topic("status");
        _mqttClient.publish(statusTopic.c_str(), "{\"status\":\"online\"}", true);

        _connected = true;
        Serial.printf("[%s] 已连接到 %s:%d\n", _deviceId, _broker, _port);
        return true;
    } else {
        Serial.printf("[%s] 连接失败, rc=%d\n", _deviceId, _mqttClient.state());
        _connected = false;
        return false;
    }
}

void IoTDevice::report(const char* jsonPayload) {
    String topic = _topic("telemetry");
    _mqttClient.publish(topic.c_str(), jsonPayload);
}

void IoTDevice::report(const String& jsonPayload) {
    report(jsonPayload.c_str());
}

void IoTDevice::report(JsonDocument& doc) {
    String payload;
    serializeJson(doc, payload);
    report(payload);
}

void IoTDevice::reportShadow(const char* jsonPayload) {
    String topic = _topic("state/reported");
    _mqttClient.publish(topic.c_str(), jsonPayload, true);
}

void IoTDevice::onCommand(CommandHandler handler) {
    _commandHandler = handler;
}

void IoTDevice::tick() {
    if (!_mqttClient.connected()) {
        _connected = false;
        _reconnect();
    }
    _mqttClient.loop();
}

bool IoTDevice::isConnected() {
    return _connected && _mqttClient.connected();
}

void IoTDevice::disconnect() {
    String statusTopic = _topic("status");
    _mqttClient.publish(statusTopic.c_str(), "{\"status\":\"offline\",\"reason\":\"manual\"}", true);
    _mqttClient.disconnect();
    _connected = false;
}

void IoTDevice::_mqttCallback(char* topic, byte* payload, unsigned int length) {
    if (_instance == nullptr || _instance->_commandHandler == nullptr) return;

    // 检查是否是命令主题
    String cmdTopic = _instance->_topic("commands");
    if (String(topic) == cmdTopic) {
        // 解析 JSON
        StaticJsonDocument<512> doc;
        DeserializationError error = deserializeJson(doc, payload, length);
        if (!error) {
            _instance->_commandHandler(doc);
        }
    }
}

void IoTDevice::_reconnect() {
    unsigned long now = millis();
    if (now - _lastReconnectAttempt < 5000) return;  // 5 秒重试间隔
    _lastReconnectAttempt = now;

    Serial.printf("[%s] 尝试重连...\n", _deviceId);
    connect();
}

String IoTDevice::_topic(const char* suffix) {
    return String("devices/") + _deviceId + "/" + suffix;
}

File: sdk/micropython/iot_device.py
"""
IoT Platform 设备接入 SDK — MicroPython 版
适用于 ESP32, ESP8266, 树莓派 Pico 等 MicroPython 设备

用法:
    from iot_device import IoTDevice

    device = IoTDevice("sensor-01", "192.168.1.100")
    device.connect()

    def on_command(cmd):
        print("收到命令:", cmd)

    device.on_command(on_command)

    while True:
        device.report({"temperature": 25.3})
        device.tick()
        time.sleep(5)
"""

import json
import time
from umqtt.simple import MQTTClient


class IoTDevice:
    """IoT 平台设备接入客户端"""

    def __init__(self, device_id, broker, port=1883):
        """
        初始化设备客户端

        Args:
            device_id: 设备 ID（需在平台预注册）
            broker: MQTT Broker 地址（平台 IP）
            port: MQTT Broker 端口（默认 1883）
        """
        self.device_id = device_id
        self.broker = broker
        self.port = port
        self.client = MQTTClient(device_id, broker, port, keepalive=60)
        self.client.set_callback(self._on_mqtt_message)
        self._command_handler = None
        self._connected = False

        # 设置 LWT（设备离线时自动发布）
        self.client.set_last_will(
            f"devices/{device_id}/status",
            json.dumps({"status": "offline", "reason": "connection_lost"}),
            retain=True, qos=1,
        )

    def connect(self):
        """连接到 MQTT Broker"""
        try:
            self.client.connect()
            self.client.subscribe(f"devices/{self.device_id}/commands")
            self.client.publish(
                f"devices/{self.device_id}/status",
                json.dumps({"status": "online"}),
                retain=True, qos=1,
            )
            self._connected = True
            print(f"[{self.device_id}] 已连接到 {self.broker}:{self.port}")
        except Exception as e:
            print(f"[{self.device_id}] 连接失败: {e}")
            self._connected = False
            raise

    def report(self, data):
        """
        上报遥测数据

        Args:
            data: dict，如 {"temperature": 25.3}
        """
        try:
            topic = f"devices/{self.device_id}/telemetry"
            self.client.publish(topic, json.dumps(data))
        except OSError:
            self._connected = False
            raise

    def report_shadow(self, state):
        """
        上报设备影子状态

        Args:
            state: dict，如 {"power": "on", "speed": 3}
        """
        try:
            topic = f"devices/{self.device_id}/state/reported"
            self.client.publish(topic, json.dumps(state))
        except OSError:
            self._connected = False
            raise

    def on_command(self, handler):
        """
        注册命令处理函数

        Args:
            handler: 回调函数，接收一个 dict 参数
        """
        self._command_handler = handler

    def tick(self):
        """
        处理 MQTT 消息（在主循环中调用）

        失败时自动重连
        """
        try:
            self.client.check_msg()
        except OSError:
            self._connected = False
            self._reconnect()

    def is_connected(self):
        """检查连接状态"""
        return self._connected

    def disconnect(self):
        """断开连接"""
        try:
            self.client.publish(
                f"devices/{self.device_id}/status",
                json.dumps({"status": "offline", "reason": "manual"}),
                retain=True, qos=1,
            )
            self.client.disconnect()
        except Exception:
            pass
        self._connected = False

    def _on_mqtt_message(self, topic, msg):
        """内部 MQTT 消息回调"""
        try:
            topic_str = topic.decode() if isinstance(topic, bytes) else topic
            if f"devices/{self.device_id}/commands" in topic_str:
                payload = json.loads(msg)
                if self._command_handler:
                    self._command_handler(payload)
        except Exception as e:
            print(f"[{self.device_id}] 消息处理异常: {e}")

    def _reconnect(self):
        """自动重连（指数退避）"""
        delay = 1
        while not self._connected:
            try:
                print(f"[{self.device_id}] {delay}秒后重连...")
                time.sleep(delay)
                self.connect()
                print(f"[{self.device_id}] 重连成功")
            except Exception:
                delay = min(delay * 2, 30)
